uefi/src/proto/shell/mod.rs

```diff
@@ -2,12 +2,208 @@
 
 /// Shell Protocol
 #[derive(Debug)]
 #[repr(transparent)]
-#[unsafe_protocol(uefi_raw::protocol::shell::ShellProtocol::GUID)]
-pub struct Shell(uefi_raw::protocol::shell::ShellProtocol);
+#[unsafe_protocol(ShellProtocol::GUID)]
+pub struct Shell(ShellProtocol);
+
+/// Iterator over the names of environmental variables obtained from the Shell protocol.
+#[derive(Debug)]
+pub struct Vars<'a> {
+    /// Char16 containing names of environment variables
+    inner: *const Char16,
+    /// Placeholder to attach a lifetime to `Vars`
+    placeholder: PhantomData<&'a CStr16>,
+}
+
+impl<'a> Iterator for Vars<'a> {
+    type Item = &'a CStr16;
+    // We iterate a list of NUL terminated CStr16s.
+    // The list is terminated with a double NUL.
+    fn next(&mut self) -> Option<Self::Item> {
+        let cur_start = self.inner;
+        let mut cur_len = 0;
+        unsafe {
+            if *(cur_start) == Char16::from_u16_unchecked(0) {
+                return None;
+            }
+            while *(cur_start.add(cur_len)) != Char16::from_u16_unchecked(0) {
+                cur_len += 1;
+            }
+            self.inner = self.inner.add(cur_len + 1);
+            Some(CStr16::from_ptr(cur_start))
+        }
+    }
+}
+
+impl Shell {
+    /// Gets the value of the specified environment variable
+    ///
+    /// # Arguments
+    ///
+    /// * `name` - The environment variable name of which to retrieve the
+    ///   value.
+    ///
+    /// # Returns
+    ///
+    /// * `Some(<env_value>)` - &CStr16 containing the value of the
+    ///   environment variable
+    /// * `None` - If environment variable does not exist
+    #[must_use]
+    pub fn get_env(&self, name: &CStr16) -> Option<&CStr16> {
+        let name_ptr: *const Char16 = core::ptr::from_ref::<CStr16>(name).cast();
+        let var_val = unsafe { (self.0.get_env)(name_ptr.cast()) };
+        if var_val.is_null() {
+            None
+        } else {
+            unsafe { Some(CStr16::from_ptr(var_val.cast())) }
+        }
+    }
+
+    /// Gets the list of environment variables
+    ///
+    /// # Returns
+    ///
+    /// * `Vec<env_names>` - Vector of environment variable names
+    #[must_use]
+    pub fn get_envs(&self) -> Vars {
+        let env_ptr = unsafe { (self.0.get_env)(ptr::null()) };
+        Vars {
+            inner: env_ptr.cast::<Char16>(),
+            placeholder: PhantomData,
+        }
+    }
+
+    /// Sets the environment variable
+    ///
+    /// # Arguments
+    ///
+    /// * `name` - The environment variable for which to set the value
+    /// * `value` - The new value of the environment variable
+    /// * `volatile` - Indicates whether the variable is volatile or
+    ///   not
+    ///
+    /// # Returns
+    ///
+    /// * `Status::SUCCESS` - The variable was successfully set
+    pub fn set_env(&self, name: &CStr16, value: &CStr16, volatile: bool) -> Status {
+        let name_ptr: *const Char16 = core::ptr::from_ref::<CStr16>(name).cast();
```

Comment:
Instead of `core::ptr::from_ref`, you can use `name.as_ptr()` (and ditto for value below).